### PR TITLE
Refactor and enhance Translatable Executor and Optimizer

### PR DESCRIPTION
**Pull request by boyjoy1127**
Ref #19937

Changes proposed in this pull request:
  - Refactor and enhance Translatable Executor and Optimizer

---

Before committing this PR, I'm sure that I have checked the following options:
- [X] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [X] I have self-reviewed the commit code.
- [X] I have (or in comment I request) added corresponding labels for the pull request.
- [X] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [X] I have made corresponding changes to the documentation.
- [X] I have added corresponding unit tests for my changes.

### CODE REVIEW

File: kernel/sql-federation/executor/core/src/main/java/org/apache/shardingsphere/sqlfederation/executor/FilterableTableScanExecutor.java [hunk 1]
```diff
@@ -117,6 +117,11 @@ public final class FilterableTableScanExecutor implements TableScanExecutor {
     
     private final EventBusContext eventBusContext;
     
+    @Override
+    public Enumerable<Object> executeScalar(final ShardingSphereTable table, final ScanNodeExecutorContext scanContext) {
+        return null;
```
Comment (strongduanmu) [inline kernel/sql-federation/executor/core/src/main/java/org/apache/shardingsphere/sqlfederation/executor/FilterableTableScanExecutor.java:122]: Why return null here? Can we return empty Enumerable?

File: kernel/sql-federation/executor/core/src/main/java/org/apache/shardingsphere/sqlfederation/executor/TranslatableTableScanExecutor.java [hunk 2]
```diff
@@ -229,26 +327,47 @@ private RelNode createRelNode(final ShardingSphereTable table, final Translatabl
...
-                result.add(StringToRexNodeUtil.buildRexNode(each, rexBuilder));
+                Map<Integer, Integer> columnMap = extractColumnMap(each);
+                String filterValue = extractFilterValue(each);
+                result.add(StringToRexNodeUtil.buildRexNode(filterValue, rexBuilder, context.getParameters(), columnMap));
             }
         }
         return result;
     }
     
+    private Map<Integer, Integer> extractColumnMap(final String filterExpression) {
```
Comment (strongduanmu) [inline kernel/sql-federation/executor/core/src/main/java/org/apache/shardingsphere/sqlfederation/executor/TranslatableTableScanExecutor.java:352]: Can you give an example of filterExpression? It looks like a json format value.

File: kernel/sql-federation/executor/core/src/main/java/org/apache/shardingsphere/sqlfederation/executor/TranslatableTableScanExecutor.java [hunk 3]
```diff
@@ -229,26 +327,47 @@ private RelNode createRelNode(final ShardingSphereTable table, final Translatabl
...
+        String columnInformationPattern = "\\{.*}";
+        Matcher matcher = Pattern.compile(columnInformationPattern).matcher(filterExpression);
```
Comment (strongduanmu) [inline kernel/sql-federation/executor/core/src/main/java/org/apache/shardingsphere/sqlfederation/executor/TranslatableTableScanExecutor.java:354]: In order to improve performance, can we extract Pattern as a static field?

File: kernel/sql-federation/optimizer/src/main/java/org/apache/shardingsphere/sqlfederation/optimizer/metadata/translatable/ParseRexNodeVisitorImpl.java [hunk 4]
```diff
@@ -138,9 +144,44 @@ public RexNode visitInput(final InputContext ctx) {
     
     @Override
     public RexNode visitInputRef(final InputRefContext ctx) {
+        // 处理inputRef位于cast的场景：CAST($1,INTEGER)。
```
Comment (strongduanmu) [inline kernel/sql-federation/optimizer/src/main/java/org/apache/shardingsphere/sqlfederation/optimizer/metadata/translatable/ParseRexNodeVisitorImpl.java:147]: Please remove inline doc here.

File: kernel/sql-federation/optimizer/src/main/java/org/apache/shardingsphere/sqlfederation/optimizer/metadata/translatable/ParseRexNodeVisitorImpl.java [hunk 5]
```diff
@@ -138,9 +144,44 @@ public RexNode visitInput(final InputContext ctx) {
...
-        RelDataType nonNullableInt = typeFactory.createSqlType(SqlTypeName.INTEGER);
-        return rexBuilder.makeInputRef(nonNullableInt, index);
+        if ((ctx.getParent() instanceof CastContext) && "VARCHAR".equals(ctx.getParent().getStop().getText())) {
+            return rexBuilder.makeInputRef(typeFactory.createJavaType(String.class), index);
+        } else if ((ctx.getParent() instanceof CastContext) && "INTEGER".equals(ctx.getParent().getStop().getText())) {
+            return rexBuilder.makeInputRef(typeFactory.createJavaType(Integer.class), index);
+        } else if ((ctx.getParent() instanceof CastContext) && "BIGINT".equals(ctx.getParent().getStop().getText())) {
+            return rexBuilder.makeInputRef(typeFactory.createJavaType(Long.class), index);
+        }
+        // 处理inputRef的普通场景：$0。
```
Comment (strongduanmu) [inline kernel/sql-federation/optimizer/src/main/java/org/apache/shardingsphere/sqlfederation/optimizer/metadata/translatable/ParseRexNodeVisitorImpl.java:156]: Please remove inline doc here.

File: kernel/sql-federation/optimizer/src/main/java/org/apache/shardingsphere/sqlfederation/optimizer/metadata/translatable/ParseRexNodeVisitorImpl.java [hunk 6]
```diff
@@ -138,9 +144,44 @@ public RexNode visitInput(final InputContext ctx) {
...
+        if (null != columnMap.get(index)) {
+            Class dataType = getClass(columnMap.get(index));
+            return rexBuilder.makeInputRef(typeFactory.createJavaType(dataType), index);
+        }
+        // 需要处理关联查询的场景：$cor0.merchant_id。
```
Comment (strongduanmu) [inline kernel/sql-federation/optimizer/src/main/java/org/apache/shardingsphere/sqlfederation/optimizer/metadata/translatable/ParseRexNodeVisitorImpl.java:161]: Please remove inline doc here.

File: kernel/sql-federation/optimizer/src/main/java/org/apache/shardingsphere/sqlfederation/optimizer/metadata/translatable/ParseRexNodeVisitorImpl.java [hunk 7]
```diff
@@ -138,9 +144,44 @@ public RexNode visitInput(final InputContext ctx) {
...
+        return rexBuilder.makeInputRef(typeFactory.createJavaType(Integer.class), index);
+    }
+    
+    private Class getClass(final int dataType) {
+        // Reference to java.sql.Types
```
Comment (strongduanmu) [inline kernel/sql-federation/optimizer/src/main/java/org/apache/shardingsphere/sqlfederation/optimizer/metadata/translatable/ParseRexNodeVisitorImpl.java:166]: Please remove inline doc here.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: kernel/sql-federation/executor/core/src/main/java/org/apache/shardingsphere/sqlfederation/executor/TranslatableTableScanExecutor.java (hunk 2 above)
Comment (boyjoy1127) [reply kernel/sql-federation/executor/core/src/main/java/org/apache/shardingsphere/sqlfederation/executor/TranslatableTableScanExecutor.java:351]: Example as : "{1=4}"
This means column 1 with type int, 4 represent int. According below:
```
private Class getClass(final int dataType) {
        switch (dataType) {
            case -5:
                return Long.class;
            case 4:
                return Integer.class;
            case 6:
                return Float.class;
            case 8:
                return Double.class;
            case 1:
                return String.class;
            case 12:
                return String.class;
            case 91:
                return Date.class;
            default:
                return String.class;
        }
```



        public final static int BIT             =  -7;

        public final static int TINYINT         =  -6;

        public final static int SMALLINT        =   5;

        public final static int INTEGER         =   4;

/**
 * <P>The constant in the Java programming language, sometimes referred
 * to as a type code, that identifies the generic SQL type
 * <code>BIGINT</code>.
 */
        public final static int BIGINT          =  -5;

/**
 * <P>The constant in the Java programming language, sometimes referred
 * to as a type code, that identifies the generic SQL type
 * <code>FLOAT</code>.
 */
        public final static int FLOAT           =   6;

/**
 * <P>The constant in the Java programming language, sometimes referred
 * to as a type code, that identifies the generic SQL type
 * <code>REAL</code>.
 */
        public final static int REAL            =   7;


/**
 * <P>The constant in the Java programming language, sometimes referred
 * to as a type code, that identifies the generic SQL type
 * <code>DOUBLE</code>.
 */
        public final static int DOUBLE          =   8;

/**
 * <P>The constant in the Java programming language, sometimes referred
 * to as a type code, that identifies the generic SQL type
 * <code>NUMERIC</code>.
 */
        public final static int NUMERIC         =   2;

/**
 * <P>The constant in the Java programming language, sometimes referred
 * to as a type code, that identifies the generic SQL type
 * <code>DECIMAL</code>.
 */
        public final static int DECIMAL         =   3;

/**
 * <P>The constant in the Java programming language, sometimes referred
 * to as a type code, that identifies the generic SQL type
 * <code>CHAR</code>.
 */
        public final static int CHAR            =   1;

/**
 * <P>The constant in the Java programming language, sometimes referred
 * to as a type code, that identifies the generic SQL type
 * <code>VARCHAR</code>.
 */
        public final static int VARCHAR         =  12;

/**
 * <P>The constant in the Java programming language, sometimes referred
 * to as a type code, that identifies the generic SQL type
 * <code>LONGVARCHAR</code>.
 */
        public final static int LONGVARCHAR     =  -1;


/**
 * <P>The constant in the Java programming language, sometimes referred
 * to as a type code, that identifies the generic SQL type
 * <code>DATE</code>.
 */
        public final static int DATE            =  91;

/**
 * <P>The constant in the Java programming language, sometimes referred
 * to as a type code, that identifies the generic SQL type
 * <code>TIME</code>.
 */
        public final static int TIME            =  92;

/**
 * <P>The constant in the Java programming language, sometimes referred
 * to as a type code, that identifies the generic SQL type
 * <code>TIMESTAMP</code>.
 */
        public final static int TIMESTAMP       =  93;


/**
 * <P>The constant in the Java programming language, sometimes referred
 * to as a type code, that identifies the generic SQL type
 * <code>BINARY</code>.
 */
        public final static int BINARY          =  -2;

/**
 * <P>The constant in the Java programming language, sometimes referred
 * to as a type code, that identifies the generic SQL type
 * <code>VARBINARY</code>.
 */
        public final static int VARBINARY       =  -3;

/**
 * <P>The constant in the Java programming language, sometimes referred
 * to as a type code, that identifies the generic SQL type
 * <code>LONGVARBINARY</code>.
 */
        public final static int LONGVARBINARY   =  -4;

/**
 * <P>The constant in the Java programming language
 * that identifies the generic SQL value
 * <code>NULL</code>.
 */
        public final static int NULL            =   0;

    /**
     * The constant in the Java programming language that indicates
     * that the SQL type is database-specific and
     * gets mapped to a Java object that can be accessed via
     * the methods <code>getObject</code> and <code>setObject</code>.
     */
        public final static int OTHER           = 1111;

**Review by strongduanmu — APPROVED** [review]
Looks great, merged.

**Pull request merged**